File: objc/objcsearch/objcsearch_tests/FileTypesTests.m
#import <XCTest/XCTest.h>
#import "FileTypes.h"

@interface FileTypesTests : XCTestCase
@property FileTypes *fileTypes;
@end

@implementation FileTypesTests

- (void)setUp {
    [super setUp];
    self.fileTypes = [[FileTypes alloc] init];
}

- (void)tearDown {
    [super tearDown];
}

- (void)testArchiveFiles {
    NSArray<NSString*> *exts = @[@"7z", @"arj", @"bz2", @"cpio", @"ear", @"gz", @"hqx", @"jar",
                                 @"pax", @"rar", @"sit", @"sitx", @"tar", @"tgz", @"war", @"zip",
                                 @"zipx", @"Z"];
    for (NSString *x in exts) {
        NSString *fileName = [NSString stringWithFormat:@"archive.%@", x];
        XCTAssert([self.fileTypes getFileType:fileName] == FileTypeArchive);
        XCTAssert([self.fileTypes isArchiveFile:fileName]);
        XCTAssert(![self.fileTypes isBinaryFile:fileName]);
        XCTAssert(![self.fileTypes isCodeFile:fileName]);
    }
}

- (void)testBinaryFiles {
    NSArray<NSString*> *exts = @[@"a", @"ai", @"beam", @"chm", @"class", @"com", @"dat",
                                 @"dll", @"doc", @"dot", @"dylib", @"exe", @"hlp", @"lib",
                                 @"mdb", @"pdb", @"so"];
    for (NSString *x in exts) {
        NSString *fileName = [NSString stringWithFormat:@"binary.%@", x];
        FileType ft = [self.fileTypes getFileType:fileName];
        // logMsg([NSString stringWithFormat:@"\nfileName: %@", fileName]);
        // logMsg([NSString stringWithFormat:@"ft: %@", [FileTypes toName:ft]]);
        // BOOL isBinary = ft == FileTypeBinary;
        // logMsg([NSString stringWithFormat:@"ft == FileTypeBinary: %hhd", isBinary]);
        XCTAssert(ft == FileTypeBinary);
        XCTAssert(![self.fileTypes isArchiveFile:fileName]);
        XCTAssert([self.fileTypes isBinaryFile:fileName]);
        XCTAssert(![self.fileTypes isCodeFile:fileName]);
    }
}

- (void)testCodeFiles {
    NSArray<NSString*> *exts = @[@"asp", @"bas", @"bash", @"bat", @"c", @"clj", @"cpp", @"cs",
                                 @"css", @"erl", @"fs", @"go", @"groovy", @"h", @"hpp", @"hs",
                                 @"htm", @"html", @"java", @"js", @"m", @"php", @"py", @"rb",
                                 @"rc", @"scala", @"sh", @"swift", @"ts", @"vb"];
    for (NSString *x in exts) {
        NSString *fileName = [NSString stringWithFormat:@"code.%@", x];
        XCTAssert([self.fileTypes getFileType:fileName] == FileTypeCode);
        XCTAssert(![self.fileTypes isArchiveFile:fileName]);
        XCTAssert(![self.fileTypes isBinaryFile:fileName]);
        XCTAssert([self.fileTypes isCodeFile:fileName]);
    }
}

- (void)testTextFiles {
    NSArray<NSString*> *exts = @[@"cfg", @"cmake", @"conf", @"csv", @"ddl", @"ics", @"ini",
                                 @"log", @"markdown", @"md", @"po", @"pot", @"properties",
                                 @"rtf", @"scc", @"sgm",  @"sgml", @"sql", @"txt", @"yml"];
    for (NSString *x in exts) {
        NSString *fileName = [NSString stringWithFormat:@"text.%@", x];
        FileType ft = [self.fileTypes getFileType:fileName];
        logMsg([NSString stringWithFormat:@"\nfileName: %@", fileName]);
        logMsg([NSString stringWithFormat:@"ft: %@", [FileTypes toName:ft]]);
        BOOL isText = ft == FileTypeText;
        logMsg([NSString stringWithFormat:@"ft == FileTypeText: %hhd", isText]);
        XCTAssert([self.fileTypes getFileType:fileName] == FileTypeText);
        XCTAssert(![self.fileTypes isArchiveFile:fileName]);
        XCTAssert(![self.fileTypes isBinaryFile:fileName]);
        XCTAssert(![self.fileTypes isCodeFile:fileName]);
        XCTAssert([self.fileTypes isTextFile:fileName]);
    }
}

@end

File: swift/swiftsearch/swiftsearchTests/FileTypesTests.swift
//
//  swiftsearchTests.swift
//  swiftsearchTests
//
//  Created by Cary Clark on 5/18/15.
//  Copyright (c) 2015 Cary Clark. All rights reserved.
//

import Cocoa
import XCTest

class FileTypesTests: XCTestCase {

    let fileTypes = FileTypes()

    override func setUp() {
        super.setUp()
    }

    override func tearDown() {
        super.tearDown()
    }

    func testArchiveFiles() {
        let exts = ["7z", "arj", "bz2", "cpio", "ear", "gz", "hqx", "jar",
            "pax", "rar", "sit", "sitx", "tar", "tgz", "war", "zip", "zipx",
            "Z"]
        for x in exts {
            let fileName = "archive.\(x)"
            XCTAssert(fileTypes.isArchiveFile(fileName), "isArchiveFile(\(fileName)) == true")
            XCTAssert(!fileTypes.isBinaryFile(fileName), "isBinaryFile(\(fileName)) == false")
            XCTAssert(!fileTypes.isTextFile(fileName), "isTextFile(\(fileName)) == false")
            XCTAssert(fileTypes.getFileType(fileName) == FileType.Archive,
                "\(fileName) is archive file")
        }
    }

    func testBinaryFiles() {
        let exts = ["a", "ai", "beam", "bin", "chm", "class", "com", "dat",
            "dbmdl", "dcr", "dir", "dll", "dms", "doc", "dot", "dxr", "dylib",
            "epub", "exe", "fm", "hi", "hlp", "indd", "lib", "lnk", "mdb", "mo",
            "mobi", "mpp", "nib", "o", "obj", "odm", "odt", "ott", "pages",
            "pdb", "ppt", "psd", "pub", "pyc", "pyo", "qxd", "rpt", "so", "swf",
            "sys", "vsd", "wpd", "wps", "wpt", "wri", "xls", "xlt"]
        for x in exts {
            let fileName = "binfile.\(x)"
            XCTAssert(!fileTypes.isArchiveFile(fileName), "isArchiveFile(\(fileName)) == false")
            XCTAssert(fileTypes.isBinaryFile(fileName), "isBinaryFile(\(fileName)) == true")
            XCTAssert(!fileTypes.isTextFile(fileName), "isTextFile(\(fileName)) == false")
            XCTAssert(fileTypes.getFileType(fileName) == FileType.Binary,
                "\(fileName) is binary file")
        }
    }

    func testGetFileTypeTextFile() {
        let fileName = "code.swift"
        XCTAssert(fileTypes.isTextFile(fileName), "isTextFile(\(fileName)) == true")
        XCTAssert(fileTypes.getFileType(fileName) == FileType.Text, "\(fileName) is binary file")
    }

    func testGetFileTypeUnknownFile() {
        let fileName = "unknown.ZZZ"
        XCTAssert(fileTypes.getFileType(fileName) == FileType.Unknown, "\(fileName) is unknown file")
    }
}
